python/pyarrow/io.pxi

```diff
@@ -65,45 +66,63 @@ cdef class NativeFile:
 
         def __get__(self):
             # Emulate built-in file modes
-            if self.is_readable and self.is_writeable:
+            if self.is_readable and self.is_writable:
                 return 'rb+'
             elif self.is_readable:
                 return 'rb'
-            elif self.is_writeable:
+            elif self.is_writable:
                 return 'wb'
             else:
                 raise ValueError('File object is malformed, has no mode')
 
+    def readable(self):
+        self._assert_open()
+        return self.is_readable
+
+    def writable(self):
+        self._assert_open()
+        return self.is_writable
+
+    def seekable(self):
+        self._assert_open()
+        return self.is_readable
+
     def close(self):
-        if self.is_open:
+        if not self.closed:
             with nogil:
```

Comment:
Unrelated, but I wonder what happens if two threads call `close()` concurrently.

Reply:
Hmmm, maybe the nogil block here should be removed?

Reply:
Well, if `rd_file.close()` or `wr_file.close()` may do I/O, you want to release the GIL.
One possible solution is something like:
```
rd_file, self.rd_file = self.rd_file, None
wr_file, self.wr_file = self.wr_file, None
with nogil:
    if self.is_readable:
        check_status(rd_file.get().close())
    else:
        check_status(rd_file.get().close())
```

Reply:
A general guideline for `close()` is that even if it fails somehow (e.g. IO error), it should still put the object in a "closed" state (and release associated resources).

Reply:
GIL should be released here -- if any Python resources are required in `Close` then the GIL will be re-acquired